### PR TITLE
Convert library functions to new specifications

### PR DESCRIPTION
**Pull request by karoliineh**
Relates to #1029.

A contribution to the effort of converting **all** library functions to new specifications.

### CODE REVIEW

File: src/analyses/libraryFunctions.ml [hunk 1]
```diff
@@ -78,14 +128,74 @@ let posix_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
+    ("nl_langinfo", unknown ~attrs:[ThreadUnsafe] [drop "item" []]);
     ("nl_langinfo_l", unknown [drop "item" []; drop "locale" [r_deep]]);
-    ("getc_unlocked", unknown [drop "stream" [w]]);
-    ("getchar_unlocked", unknown []);
-    ("putc_unlocked", unknown [drop "c" []; drop "stream" [w]]);
-    ("putchar_unlocked", unknown [drop "c" []]);
+    ("getc_unlocked", unknown ~attrs:[ThreadUnsafe] [drop "stream" [r_deep; w_deep]]);
+    ("getchar_unlocked", unknown ~attrs:[ThreadUnsafe] []);
+    ("ptsname", unknown ~attrs:[ThreadUnsafe] [drop "fd" []]);
+    ("putc_unlocked", unknown ~attrs:[ThreadUnsafe] [drop "c" []; drop "stream" [w]]);
```
Comment (sim642) [inline src/analyses/libraryFunctions.ml:183]:
```suggestion
    ("putc_unlocked", unknown ~attrs:[ThreadUnsafe] [drop "c" []; drop "stream" [r_deep; w_deep]]);
```

File: src/analyses/libraryFunctions.ml [hunk 2]
```diff
@@ -78,14 +128,74 @@ let posix_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
+    ("putchar_unlocked", unknown ~attrs:[ThreadUnsafe] [drop "c" []]);
+    ("putenv", unknown ~attrs:[ThreadUnsafe] [drop "string" [r; w]]);
+    ("readdir", unknown ~attrs:[ThreadUnsafe] [drop "dirp" [r_deep]]);
+    ("setenv", unknown ~attrs:[ThreadUnsafe] [drop "name" [r]; drop "name" [r]; drop "overwrite" []]);
+    ("setgrent", unknown ~attrs:[ThreadUnsafe] []);
+    ("setpwent", unknown ~attrs:[ThreadUnsafe] []);
+    ("setutxent", unknown ~attrs:[ThreadUnsafe] []);
+    ("strerror", unknown ~attrs:[ThreadUnsafe] [drop "errnum" []]);
```
Comment (sim642) [inline src/analyses/libraryFunctions.ml:191]: Should be in c group.

File: src/analyses/libraryFunctions.ml [hunk 3]
```diff
@@ -78,14 +128,74 @@ let posix_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
+    ("strsignal", unknown ~attrs:[ThreadUnsafe] [drop "sig" []]);
+    ("unsetenv", unknown ~attrs:[ThreadUnsafe] [drop "name" [r]]);
     ("lseek", unknown [drop "fd" []; drop "offset" []; drop "whence" []]);
-    ("fseeko", unknown [drop "stream" [w]; drop "offset" []; drop "whence" []]);
+    ("fcntl", unknown (drop "fd" [] :: drop "cmd" [] :: VarArgs (drop' [r; w])));
+    ("fseeko", unknown [drop "stream" [r_deep; w_deep]; drop "offset" []; drop "whence" []]);
+    ("fileno", unknown [drop "stream" [r_deep; w_deep]]);
+    ("getopt", unknown ~attrs:[ThreadUnsafe] [drop "argc" []; drop "argv" [r]; drop "optstring" [r]]);
```
Comment (sim642) [inline src/analyses/libraryFunctions.ml:198]:
```suggestion
    ("getopt", unknown ~attrs:[ThreadUnsafe] [drop "argc" []; drop "argv" [r_deep]; drop "optstring" [r]]);
```

File: src/analyses/libraryFunctions.ml [hunk 4]
```diff
@@ -128,7 +240,22 @@ let posix_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
     ("timer_getoverrun", unknown [drop "timerid" []]);
     ("lstat", unknown [drop "pathname" [r]; drop "statbuf" [w]]);
     ("getpwnam", unknown [drop "name" [r]]);
+    ("chdir", unknown [drop "path" [r]]);
+    ("closedir", unknown [drop "dirp" [w]]);
```
Comment (sim642) [inline src/analyses/libraryFunctions.ml:244]:
```suggestion
    ("closedir", unknown [drop "dirp" [r]]);
```

File: src/analyses/libraryFunctions.ml [hunk 5]
```diff
@@ -158,15 +286,55 @@ let posix_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
+    ("pthread_rwlock_init", unknown [drop "rwlock" [w]; drop "attr" [w]]);
+    ("pthread_rwlock_destroy", unknown [drop "rwlock" [f]]);
+    ("pthread_rwlock_rdlock", special [__ "rwlock" []] @@ fun rwlock -> Lock {lock = rwlock; try_ = get_bool "sem.lock.fail"; write = false; return_on_success = true});
+    ("pthread_rwlock_tryrdlock", special [__ "rwlock" []] @@ fun rwlock -> Lock {lock = rwlock; try_ = get_bool "sem.lock.fail"; write = false; return_on_success = true});
+    ("pthread_rwlock_wrlock", special [__ "rwlock" []] @@ fun rwlock -> Lock {lock = rwlock; try_ = get_bool "sem.lock.fail"; write = true; return_on_success = true});
+    ("pthread_rwlock_trywrlock", special [__ "rwlock" []] @@ fun rwlock -> Lock {lock = rwlock; try_ = true; write = true; return_on_success = false});
+    ("pthread_rwlock_unlock", special [__ "rwlock" []] @@ fun rwlock -> Unlock rwlock);
+    ("pthread_rwlockattr_init", unknown [drop "attr" [w]]);
+    ("pthread_rwlockattr_destroy", unknown [drop "attr" [f]]);
+    ("pthread_spin_init", unknown [drop "lock" []; drop "pshared" []]);
```
Comment (sim642) [inline src/analyses/libraryFunctions.ml:320]:
```suggestion
    ("pthread_spin_init", unknown [drop "lock" [w]; drop "pshared" []]);
```

File: src/analyses/libraryFunctions.ml [hunk 6]
```diff
@@ -158,15 +286,55 @@ let posix_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
 let pthread_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
     ("pthread_create", special [__ "thread" [w]; drop "attr" [r]; __ "start_routine" [s]; __ "arg" []] @@ fun thread start_routine arg -> ThreadCreate { thread; start_routine; arg }); (* For precision purposes arg is not considered accessed here. Instead all accesses (if any) come from actually analyzing start_routine. *)
     ("pthread_exit", special [__ "retval" []] @@ fun retval -> ThreadExit { ret_val = retval }); (* Doesn't dereference the void* itself, but just passes to pthread_join. *)
+    ("pthread_cond_init", unknown [drop "cond" [w]; drop "attr" [w]]);
+    ("__pthread_cond_init", unknown [drop "cond" [w]; drop "attr" [w]]);
```
Comment (sim642) [inline src/analyses/libraryFunctions.ml:290]:
```suggestion
    ("pthread_cond_init", unknown [drop "cond" [w]; drop "attr" [r]]);
    ("__pthread_cond_init", unknown [drop "cond" [w]; drop "attr" [r]]);
```

File: src/analyses/libraryFunctions.ml [hunk 7]
```diff
@@ -158,15 +286,55 @@ let posix_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
+    ("pthread_mutex_destroy", unknown [drop "mutex" [f]]);
+    ("pthread_mutex_lock", special [__ "mutex" []] @@ fun mutex -> Lock {lock = mutex; try_ = get_bool "sem.lock.fail"; write = true; return_on_success = false});
+    ("__pthread_mutex_lock", special [__ "mutex" []] @@ fun mutex -> Lock {lock = mutex; try_ = get_bool "sem.lock.fail"; write = true; return_on_success = false});
+    ("pthread_mutex_trylock", special [__ "mutex" []] @@ fun mutex -> Lock {lock = mutex; try_ = true; write = true; return_on_success = false});
+    ("__pthread_mutex_trylock", special [__ "mutex" []] @@ fun mutex -> Lock {lock = mutex; try_ = true; write = true; return_on_success = false});
+    ("pthread_mutex_unlock", special [__ "mutex" []] @@ fun mutex -> Unlock mutex);
+    ("__pthread_mutex_unlock", special [__ "mutex" []] @@ fun mutex -> Unlock mutex);
+    ("pthread_mutexattr_init", unknown [drop "attr" [w]]);
+    ("pthread_mutexattr_destroy", unknown [drop "attr" [f]]);
+    ("pthread_rwlock_init", unknown [drop "rwlock" [w]; drop "attr" [w]]);
```
Comment (sim642) [inline src/analyses/libraryFunctions.ml:311]:
```suggestion
    ("pthread_rwlock_init", unknown [drop "rwlock" [w]; drop "attr" [r]]);
```

File: src/analyses/libraryFunctions.ml [hunk 8]
```diff
@@ -158,15 +286,55 @@ let posix_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
+    ("pthread_spin_trylock", special [__ "lock" []] @@ fun lock -> Lock {lock = lock; try_ = true; write = true; return_on_success = false});
+    ("pthread_spin_unlock", special [__ "lock" []] @@ fun lock -> Unlock lock);
+    ("pthread_attr_init", unknown [drop "attr" [w]]);
     ("pthread_attr_destroy", unknown [drop "attr" [f]]);
+    ("pthread_attr_getdetachstate", unknown [drop "attr" [r]; drop "detachstate" [r]]);
+    ("pthread_attr_setdetachstate", unknown [drop "attr" [w]; drop "detachstate" []]);
+    ("pthread_attr_getstacksize", unknown [drop "attr" [r]; drop "stacksize" [r]]);
+    ("pthread_attr_setstacksize", unknown [drop "attr" [w]; drop "stacksize" []]);
+    ("pthread_attr_getscope", unknown [drop "attr" [r]; drop "scope" [r]]);
+    ("pthread_attr_setscope", unknown [drop "attr" [w]; drop "scope" []]);
```
Comment (sim642) [inline src/analyses/libraryFunctions.ml:332]:
```suggestion
    ("pthread_attr_getdetachstate", unknown [drop "attr" [r]; drop "detachstate" [w]]);
    ("pthread_attr_setdetachstate", unknown [drop "attr" [w]; drop "detachstate" []]);
    ("pthread_attr_getstacksize", unknown [drop "attr" [r]; drop "stacksize" [w]]);
    ("pthread_attr_setstacksize", unknown [drop "attr" [w]; drop "stacksize" []]);
    ("pthread_attr_getscope", unknown [drop "attr" [r]; drop "scope" [w]]);
    ("pthread_attr_setscope", unknown [drop "attr" [w]; drop "scope" []]);
```

File: src/analyses/libraryFunctions.ml [hunk 9]
```diff
@@ -226,12 +402,16 @@ let gcc_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
     ("fputs_unlocked", unknown [drop "s" [r]; drop "stream" [w]]);
-    ("futimesat", unknown [drop "dirfd" [w]; drop "pathname" [r]; drop "times" [r]]);
+    ("futimesat", unknown [drop "dirfd" []; drop "pathname" [r]; drop "times" [r]]);
     ("error", unknown ((drop "status" []):: (drop "errnum" []) :: (drop "format" [r]) :: (VarArgs (drop' [r]))));
     ("gettext", unknown [drop "msgid" [r]]);
     ("euidaccess", unknown [drop "pathname" [r]; drop "mode" []]);
     ("rpmatch", unknown [drop "response" [r]]);
     ("getpagesize", unknown []);
+    ("__fgets_alias", unknown [drop "__s" [r; w]; drop "__n" []; drop "__stream" [r_deep; w_deep]]);
+    ("__fgets_chk", unknown [drop "__s" [r; w]; drop "__size" []; drop "__n" []; drop "__stream" [r_deep; w_deep]]);
```
Comment (sim642) [inline src/analyses/libraryFunctions.ml:412]:
```suggestion
    ("__fgets_alias", unknown [drop "__s" [w]; drop "__n" []; drop "__stream" [r_deep; w_deep]]);
    ("__fgets_chk", unknown [drop "__s" [w]; drop "__size" []; drop "__n" []; drop "__stream" [r_deep; w_deep]]);
```

File: src/analyses/libraryFunctions.ml [hunk 10]
```diff
@@ -226,12 +402,16 @@ let gcc_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
+    ("__fread_alias", unknown [drop "__ptr" [w_deep]; drop "__size" []; drop "__n" []; drop "__stream" [r_deep; w_deep]]);
+    ("__fread_chk", unknown [drop "__ptr" [w_deep]; drop "__ptrlen" []; drop "__size" []; drop "__n" []; drop "__stream" [r_deep; w_deep]]);
```
Comment (sim642) [inline src/analyses/libraryFunctions.ml:414]:
```suggestion
    ("__fread_alias", unknown [drop "__ptr" [w]; drop "__size" []; drop "__n" []; drop "__stream" [r_deep; w_deep]]);
    ("__fread_chk", unknown [drop "__ptr" [w]; drop "__ptrlen" []; drop "__size" []; drop "__n" []; drop "__stream" [r_deep; w_deep]]);
```

File: src/analyses/libraryFunctions.ml [hunk 11]
```diff
@@ -257,16 +437,19 @@ let glibc_desc_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
 let linux_userspace_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
     (* ("prctl", unknown [drop "option" []; drop "arg2" []; drop "arg3" []; drop "arg4" []; drop "arg5" []]); *)
     ("prctl", unknown (drop "option" [] :: VarArgs (drop' []))); (* man page has 5 arguments, but header has varargs and real-world programs may call with <5 *)
     ("__ctype_tolower_loc", unknown []);
     ("__ctype_toupper_loc", unknown []);
+    ("endutxent", unknown ~attrs:[ThreadUnsafe] []);
     ("epoll_create", unknown [drop "size" []]);
     ("epoll_ctl", unknown [drop "epfd" []; drop "op" []; drop "fd" []; drop "event" [w]]);
     ("epoll_wait", unknown [drop "epfd" []; drop "events" [w]; drop "maxevents" []; drop "timeout" []]);
+    ("__fprintf_chk", unknown [drop "stream" [r_deep; w_deep]; drop "flag" []; drop "format" [r]]);
```
Comment (sim642) [inline src/analyses/libraryFunctions.ml:452]: This should also have `VarArgs` like `fprintf`.

File: src/analyses/libraryFunctions.ml [hunk 12]
```diff
@@ -78,14 +128,74 @@ let posix_descs_list: (string * LibraryDesc.t) list = LibraryDsl.[
...
+    ("dbm_delete", unknown ~attrs:[ThreadUnsafe] [drop "db" [r_deep; w_deep]; drop "key" []]);
+    ("dbm_error", unknown ~attrs:[ThreadUnsafe] [drop "db" [r_deep]]);
+    ("dbm_fetch", unknown ~attrs:[ThreadUnsafe] [drop "db" [r_deep]; drop "key" []]);
+    ("dbm_firstkey", unknown ~attrs:[ThreadUnsafe] [drop "db" [r_deep]]);
+    ("dbm_nextkey", unknown ~attrs:[ThreadUnsafe] [drop "db" [r_deep]]);
+    ("dbm_open", unknown ~attrs:[ThreadUnsafe] [drop "file" [r_deep; w_deep]; drop "open_flags" []; drop "file_mode" []]);
+    ("dbm_store", unknown ~attrs:[ThreadUnsafe] [drop "db" [r_deep; w_deep]; drop "key" []; drop "content" []; drop "store_mode" []]);
+    ("dlerror", unknown ~attrs:[ThreadUnsafe] []);
+    ("drand48", unknown ~attrs:[ThreadUnsafe] []);
+    ("encrypt", unknown ~attrs:[ThreadUnsafe] [drop "block" []; drop "edflag" []]);
```
[the remark this answers is not in the document]
Comment (karoliineh) [reply src/analyses/libraryFunctions.ml:145]: Why though? `block` is a char, not a pointer.

**Review by sim642 — APPROVED** [review]

**Pull request merged**